sharktank/sharktank/ops/utils.py

```diff
@@ -0,0 +1,175 @@
+
+    shard_count = first_replicated_tensor_arg.shard_count
+
+    per_shard_args_and_kwargs = [
+        _extract_per_shard_args_and_kwargs(i, args, kwargs) for i in range(shard_count)
+    ]
+    per_shard_results = [
+        fn(*per_shard_args, **per_shard_kwargs)
+        for per_shard_args, per_shard_kwargs in per_shard_args_and_kwargs
+    ]
+
+    def reduce_to_list_of_tensors_fn(a: Any, b: Any) -> Any:
+        if not is_any_tensor(b):
+            return a
+        a.append(b)
+        return a
+
+    def tensor_to_empty_list(x: Any) -> list:
+        if is_any_tensor(x):
+            return []
+        return x
+
+    reduce_initial = tree.map_leaves(
+        per_shard_results[0], f=tensor_to_empty_list, is_leaf=is_leaf
+    )
+
+    # Make a lists of tensor shards that correspond to the same replicated tensor.
+    reduced_results = tree.reduce_horizontal(
+        fn=reduce_to_list_of_tensors_fn,
+        trees=per_shard_results,
+        is_leaf=is_leaf,
+        initial=reduce_initial,
+    )
+
+    def make_replicated_tensor_if_tensor_collection(x: Any) -> Any:
+        if _is_tensor_collection(x):
+            return ReplicatedTensor(ts=x, devices=first_replicated_tensor_arg.devices)
+        return x
+
+    result_with_replicated_tensor = tree.map_leaves(
+        reduced_results,
+        f=make_replicated_tensor_if_tensor_collection,
+        is_leaf=_is_leaf_or_tensor_collection,
+    )
+    return result_with_replicated_tensor
+
+
+def trivially_replicable(fn: Callable[..., Any]) -> Callable[..., Any]:
+    """A decorator to turn an trivially replicable operation into a replicated
+    operation.
+
+    An operation is trivially replicable if the replicated variant just executes the
+    operation on all devices with the respective shards.
+
+    E.g.
+    ```
+    @trivially_replicable
+    def fn(a: torch.Tensor) -> torch.Tensor:
+        return a
+
+    arg = torch.Tensor([1, 2, 3])
+    shard_count = 2
+    replicated_arg = ReplicatedTensor(
+        ts=arg, shard_count=shard_count
+    )
+    fn(replicated_arg)
+    ```
+
+    ```
+    ReplicatedTensor(<unnamed>, [3], shard_count=2 of [3])
+    ```
+    """
+
+    def wrapper(*args, **kwargs):
+        return call_trivially_replicable(fn, args, kwargs)
+
+    return wrapper
+
+
+def _extract_per_shard_args_and_kwargs(
+    shard_index: int, args: tuple[Any, ...], kwargs: dict[str, Any]
+) -> tuple[tuple[Any, ...], dict[str, Any]]:
+    return tree.map_leaves(
+        (args, kwargs),
+        f=functools.partial(
+            _extract_tensor_shard_if_replicate, shard_index=shard_index
+        ),
+        is_leaf=_is_leaf,
+    )
+
+
+def _extract_tensor_shard_if_replicate(v: Any, shard_index: int) -> Any:
+    from sharktank.types import (
+        ReplicatedTensor,
+    )
+
+    if isinstance(v, ReplicatedTensor):
+        return v.shards[shard_index]
+    return v
+
+
+def _is_leaf(v: Any) -> bool:
+    from sharktank.types import (
+        is_any_tensor,
+    )
```

Comment:
Is the comma after `is_any_tensor` needed? We don't do this with other single imports, and it's causing black to format this over 3 lines instead of 1, making both this and the outer function much longer than necessary. Same comment for many of the other one-line functions